app/auth/views/proton.py

```diff
@@ -0,0 +1,119 @@
+from app.proton.proton_callback_handler import ProtonCallbackHandler, Action
+from app.utils import encode_url
+
+_authorization_base_url = PROTON_BASE_URL + "/oauth/authorize"
+_token_url = PROTON_BASE_URL + "/oauth/token"
+
+# need to set explicitly redirect_uri instead of leaving the lib to pre-fill redirect_uri
+# when served behind nginx, the redirect_uri is localhost... and not the real url
+_redirect_uri = URL + "/auth/proton/callback"
+
+
+def get_client_ip() -> Optional[str]:
+    if request.environ.get("HTTP_X_FORWARDED_FOR") is None:
+        return request.environ.get("REMOTE_ADDR")
+    else:
+        return request.environ["HTTP_X_FORWARDED_FOR"]  # if behind a proxy
+
+
+def extract_action() -> Action:
+    action = request.args.get("action")
+    if action is not None:
+        if action == "link":
+            return Action.Link
+        else:
+            raise Exception(f"Unknown action: {action}")
+    return Action.Login
+
+
+def get_action_from_state() -> Action:
+    oauth_flow_state = json.loads(session["oauth_flow_state"])
+    oauth_action = oauth_flow_state.get("action")
+    if oauth_action == Action.Login.value:
+        return Action.Login
+    elif oauth_action == Action.Link.value:
+        return Action.Link
+    raise Exception(f"Unknown action in state: {oauth_action}")
+
+
+@auth_bp.route("/proton/login")
+def proton_login():
+    if PROTON_CLIENT_ID is None or PROTON_CLIENT_SECRET is None:
+        return redirect(url_for("auth.login"))
+
+    next_url = request.args.get("next")
+    if next_url:
+        redirect_uri = _redirect_uri + "?next=" + encode_url(next_url)
+    else:
+        redirect_uri = _redirect_uri
+
+    proton = OAuth2Session(PROTON_CLIENT_ID, redirect_uri=redirect_uri)
+    authorization_url, state = proton.authorization_url(_authorization_base_url)
+
+    # State is used to prevent CSRF, keep this for later.
+    session["oauth_state"] = state
+    session["oauth_flow_state"] = json.dumps({"action": extract_action().value})
+    return redirect(authorization_url)
+
+
+@auth_bp.route("/proton/callback")
+def proton_callback():
+    if PROTON_CLIENT_ID is None or PROTON_CLIENT_SECRET is None:
+        return redirect(url_for("auth.login"))
+
+    # user clicks on cancel
+    if "error" in request.args:
+        flash("Please use another sign in method then", "warning")
+        return redirect("/")
+
+    proton = OAuth2Session(
+        PROTON_CLIENT_ID,
+        state=session["oauth_state"],
+        redirect_uri=_redirect_uri,
+    )
+    token = proton.fetch_token(
+        _token_url,
+        client_secret=PROTON_CLIENT_SECRET,
+        authorization_response=request.url,
+        verify=PROTON_VALIDATE_CERTS,
+        method="GET",
+        include_client_id=True,
+    )
+    credentials = convert_access_token(token["access_token"])
+    action = get_action_from_state()
+
+    client_original_ip = get_client_ip()
+    proton_client = HttpProtonClient(
```

Comment:
I wonder if we can use a Procedural approach here instead (i.e. not based on classes)? I feel the logic can be better followed and it is the current approach used in the repo. 

For example, we can have a method that handles the "login" action and another for the "link" action. All the logic can happen in these methods, eventually with some helper method. We can always isolate and test the helper method without using dependency injection. What do you think?

Reply:
I would keep it as it is to start homogenizing it with the rest of the proton codebase. Going back and forth between sl and proton gets confusing. In the rest of the codebase we use interfaces a lot. Afaik this is the only way to implement interfaces in python and have the type checks pass.

Reply:
It's not php or js pattern, having interfaces is a quite common way of having extensible logic in almost all languages. It's not used in the backend but in all the big codebases in proton. backend, web, ios, android... we have this as a common way of doing things.

Even if switching codebases is less frequent, its good that they share a way of doing things. This allows people to read and follow other codes consistently.

Reply:
I tried to read the code again and it's really hard to follow the logic as the call stack is deep. I think a code should be optimised for simplicity as it'd allow greater flexibility and maintainability in the future. 

The debate about object oriented vs procedural is actually useless. A procedural code can be also be as complex as object-oriented one if done wrong.

At some point, it's up to the engineering team to maintain the code base and I won't have the time to do code review so the decision of what pattern used is up to you.